Engineering/Atomic package in Golang.md

```diff
@@ -0,0 +1,138 @@
+---
+tags: engineering, golang, concurrency, race-condition
+author: Hieu Phan
+---
+
+We often run some functions asynchronously in real projects using the go routine. The problem we're facing is race-condition when updating the same variable. The solutions can be using a mutex lock or concurrency patterns to change our situation using the channels. However, In this document, we want to solve this problem when we want to update the same memory resource. Go standard library provide `sync/atomic` package to solve our problem.
+
+## Problem
+
+We need to add a value to an integer value. It's elementary logic until we run the logic many times and asynchronously. Run three times with a loop of 2000; our expectation is 6000. However, the result is not stable: many times with values less than 6000.
+```go
+package main
+
+import (
+	"fmt"
+	"sync"
+)
+
+func main() {
+	var i int32
+	var wg sync.WaitGroup
+	
+	wg.Add(3)
+	
+	go Process(&i, &wg)
+	go Process(&i, &wg)
+	go Process(&i, &wg)
+	
+	wg.Wait()
+	fmt.Println("i:", i)
+}
+
+func Process(variable *int32, wg *sync.WaitGroup) {
+	defer wg.Done()
+	
+	for i := 0; i < 2000; i++ {
+		*variable++ // The race condition makes the result NOT always equal 6000
+	}
+}
+
+```
+
+## Solutions
+
+### Using Mutex
+We run three go routines in the above example code to update the same `i` variable. Our expectation: The value always equals 6000. However, we got the race condition. The simple solution is to pass a `mutex` lock to update the value synchronously.
+
+```go
+func Process(variable *int32, wg *sync.WaitGroup, mu *sync.Mutex) {
+	defer wg.Done()
+	for i := 0; i < 2000; i++ {
+		mu.Lock()
+		*variable++
+		mu.Unlock()
+	}
+}
+```
+
+### Using Atomic
+
+Package `sync/atomic` offers primitives for atomic memory that are low-level and useful for implementing synchronization algorithms. It encapsulates the synchronous logic in the utility functions.
+
+```go
+func Process(variable *int32, wg *sync.WaitGroup) {
+	defer wg.Done()
+	for i := 0; i < 2000; i++ {
+		atomic.AddInt32(variable, 1)
+	}
+}
+```
+
+### Benchmark solutions
+
+I make a benchmark test for three implementations. The logic using `atomic` quite fast than `mutex lock`, around 0.33 % in this case.
+
+| Benchmark                         | Run   | Speed           |
+| --------------------------------- | ----- | --------------- |
+| BenchmarkAddValueRacedCondition-8 | 93110 | 12423 ns/op     |
+| BenchmarkAddValueWithMutex-8      | 10000 | 126994 ns/op    |
+| BenchmarkAddValueWithAtomic-8     | 14080 | **84896 ns/op** |
+
+In the meantime, this package support functions to interact with some types in golang: int32, int64, uint32, uint64, and the pointer. The implementation with Pointer is an excellent feature. It can get easier to apply the help of the atomic package for other types. It provides an interface to store, update, and retrieve a value of a specific type and is asynchronously included.
```

Comment:
```suggestion
In the meantime, this package support functions to interact with some types in Golang: int32, int64, uint32, uint64, and the pointer. The implementation with Pointer is an excellent feature. It can get easier to apply the help of the atomic package for other types. It provides an interface to store, update, and retrieve a value of a specific type and is asynchronously included.
```